piet-wgsl/src/main.rs

```diff
@@ -54,23 +76,186 @@ async fn do_render(
     queue: &Queue,
     engine: &mut Engine,
 ) -> Result<(), Box<dyn std::error::Error>> {
-    let shaders = shaders::init_shaders(device, engine)?;
-    let scene = test_scene::gen_test_scene();
-    dump_scene_info(&scene);
-    let (recording, buf) = render(&scene, &shaders);
-    let downloads = engine.run_recording(&device, &queue, &recording)?;
+    let (downloads, buf) =
+        render_demo_scene(device, queue, engine, DownloadBufUsage::MapRead).await?;
     let mapped = downloads.map();
     device.poll(wgpu::Maintain::Wait);
     let buf = mapped.get_mapped(buf).await?;
 
     let file = File::create("image.png")?;
     let w = BufWriter::new(file);
-    let encoder = png::Encoder::new(w, 1024, 1024);
+    let encoder = png::Encoder::new(w, WIDTH, HEIGHT);
     let mut writer = encoder.write_header()?;
     writer.write_image_data(&buf)?;
     Ok(())
 }
 
+// ------ WASM ------
+struct BufferDimensions {
+    width: u32,
+    height: u32,
+    unpadded_bytes_per_row: u32,
+    padded_bytes_per_row: u32,
+}
+
+impl BufferDimensions {
+    fn new(width: u32, height: u32) -> Self {
+        let bytes_per_pixel = std::mem::size_of::<u32>() as u32;
+        let unpadded_bytes_per_row = width * bytes_per_pixel;
+        let align = wgpu::COPY_BYTES_PER_ROW_ALIGNMENT;
+        let alignment_mask = align - 1;
+        let padded_bytes_per_row = (unpadded_bytes_per_row + alignment_mask) & !alignment_mask;
+        Self {
+            width,
+            height,
+            unpadded_bytes_per_row,
+            padded_bytes_per_row,
+        }
+    }
+
+    fn get_extent(&self) -> wgpu::Extent3d {
+        let bytes_per_pixel = std::mem::size_of::<u32>() as u32;
+        wgpu::Extent3d {
+            width: self.width,
+            height: self.height,
+            depth_or_array_layers: 1,
+        }
+    }
+}
+
+async fn run_wasm(event_loop: EventLoop<()>, window: Window) {
```

Comment:
It's not clear to me why the winit usage is locked to wasm.

Additionally, perhaps this should just be an example?